components/layout_2020/display_list/stacking_context.rs

```diff
@@ -500,63 +581,154 @@ impl StackingContext {
+        // Step 5: Float stacking containers
+        for (i, child) in self.float_stacking_containers.iter().enumerate() {
+            self.debug_push_print_item(DebugPrintField::FloatStackingContainers, i);
+            child.build_display_list(builder);
         }
 
-        // Step 6: Content
-        while child_fragments.peek().map_or(false, |child| {
-            child.section == StackingContextSection::Content
+        // Steps 6 and 7: Fragments and inline stacking containers
+        while contents.peek().map_or(false, |(_, child)| {
+            child.section() == StackingContextSection::Foreground
         }) {
-            child_fragments.next().unwrap().build_display_list(builder);
+            let (i, child) = contents.next().unwrap();
+            self.debug_push_print_item(DebugPrintField::Contents, i);
+            child.build_display_list(builder, &self.atomic_inline_stacking_containers);
         }
 
-        // Step 7, 8 & 9: Inlines that generate stacking contexts and positioned
-        // descendants with nonnegative, numeric z-indices
-        for child_context in child_stacking_contexts {
-            child_context.build_display_list(builder);
+        // Steps 8 and 9: Stacking contexts with non-negative ‘z-index’, and
+        // positioned stacking containers (where ‘z-index’ is auto)
+        for (i, child) in real_stacking_contexts_and_positioned_stacking_containers {
+            self.debug_push_print_item(
+                DebugPrintField::RealStackingContextsAndPositionedStackingContainers,
+                i,
+            );
+            child.build_display_list(builder);
         }
 
         // Step 10: Outline
-        while child_fragments.peek().map_or(false, |child| {
-            child.section == StackingContextSection::Outline
+        while contents.peek().map_or(false, |(_, child)| {
+            child.section() == StackingContextSection::Outline
         }) {
-            child_fragments.next().unwrap().build_display_list(builder);
+            let (i, child) = contents.next().unwrap();
+            self.debug_push_print_item(DebugPrintField::Contents, i);
+            child.build_display_list(builder, &self.atomic_inline_stacking_containers);
         }
 
         if pushed_context {
             builder.display_list.wr.pop_stacking_context();
         }
     }
+
+    /// Store the fact that something was painted, if [Self::debug_print_items] is not None.
+    ///
+    /// This is used to help reconstruct the original painting order in [Self::debug_print] without
+    /// duplicating our painting order logic, since that could fall out of sync with the real logic.
+    fn debug_push_print_item(&self, field: DebugPrintField, index: usize) {
+        if let Some(items) = self.debug_print_items.as_ref() {
+            items.borrow_mut().push(DebugPrintItem { field, index });
+        }
+    }
+
+    /// Print the stacking context tree, or panics if [Self::debug_print_items] is None.
```

Comment:
Looks like this doesn't panic. I think that's a better behavior actually, but perhaps the comment should be corrected.

Reply:
done in 43692821f9bccde5eda1b5651e21c535bfaebe9d